File: transfers/fraud/typologies/invoice.py
from datetime import timedelta

from common.channels import FRAUD_INVOICE
from common.transactions import Transaction
from transfers.factory import TransactionDraft

from ..engine import IllicitContext
from ..rings import Plan
from .common import append_bounded_txn


def generate(
    ctx: IllicitContext,
    ring_plan: Plan,
    budget: int,
) -> list[Transaction]:
    if budget <= 0:
        return []
    if not ring_plan.fraud_accounts or not ctx.biller_accounts:
        return []

    rng = ctx.execution.rng
    base = ctx.window.start_date + timedelta(
        days=rng.int(0, max(1, ctx.window.days - 14))
    )
    weeks = max(1, min(6, ctx.window.days // 7))

    out: list[Transaction] = []
    for _ in range(rng.int(3, 10)):
        if len(out) >= budget:
            break

        src = rng.choice(ring_plan.fraud_accounts)
        dst = rng.choice(ctx.biller_accounts)

        ln = float(rng.gen.lognormal(mean=8.0, sigma=0.35))
        amt = round(ln / 10.0) * 10.0

        ts = base + timedelta(
            days=7 * rng.int(0, weeks),
            hours=rng.int(9, 18),
            minutes=rng.int(0, 60),
        )
        if not append_bounded_txn(
            ctx,
            out,
            budget,
            TransactionDraft(
                source=src,
                destination=dst,
                amount=amt,
                timestamp=ts,
                is_fraud=1,
                ring_id=ring_plan.ring_id,
                channel=FRAUD_INVOICE,
            ),
        ):
            break

    return out

File: src/transfers/fraud/typologies/invoice.cpp
#include "phantomledger/transfers/fraud/typologies/invoice.hpp"

#include "phantomledger/probability/distributions/lognormal.hpp"
#include "phantomledger/taxonomies/channels/types.hpp"
#include "phantomledger/transactions/draft.hpp"
#include "phantomledger/transfers/fraud/typologies/common.hpp"

#include <algorithm>
#include <cmath>

namespace PhantomLedger::transfers::fraud::typologies::invoice {

std::vector<transactions::Transaction>
generate(IllicitContext &ctx, const Plan &plan, std::int32_t budget) {
  std::vector<transactions::Transaction> out;
  if (budget <= 0) {
    return out;
  }
  if (plan.fraudAccounts.empty() || ctx.billerAccounts.empty()) {
    return out;
  }

  random::Rng &rng = *ctx.execution.rng;

  const auto baseRange = std::max<std::int32_t>(1, ctx.window.days - 14);
  const auto baseOffsetDays = static_cast<std::int32_t>(
      rng.uniformInt(0, static_cast<std::int64_t>(baseRange) + 1));
  const auto baseDate = ctx.window.startDate + time::Days{baseOffsetDays};

  // weeks = max(1, min(6, days // 7))
  const auto weeks =
      std::max<std::int32_t>(1, std::min<std::int32_t>(6, ctx.window.days / 7));

  // Number of invoice events in this ring: rng.int(3, 10) → [3, 11).
  const auto events = static_cast<std::int32_t>(rng.uniformInt(3, 11));

  const auto channel = channels::tag(channels::Fraud::invoice);

  for (std::int32_t i = 0; i < events; ++i) {
    if (static_cast<std::int32_t>(out.size()) >= budget) {
      break;
    }

    const auto &src = typologies::pickOne(rng, plan.fraudAccounts);
    const auto &dst = typologies::pickOne(rng, ctx.billerAccounts);

    const double raw = probability::distributions::lognormal(rng, 8.0, 0.35);
    const double amount = std::round(raw / 10.0) * 10.0;

    // Day-of-burst: rng.int(0, weeks) inclusive → [0, weeks + 1).
    const auto weekIdx = static_cast<std::int32_t>(
        rng.uniformInt(0, static_cast<std::int64_t>(weeks) + 1));
    // Hour: rng.int(9, 18) → [9, 19); minute: rng.int(0, 60) → [0, 61).
    const auto hour = static_cast<std::int32_t>(rng.uniformInt(9, 19));
    const auto minute = static_cast<std::int32_t>(rng.uniformInt(0, 61));

    const auto ts = baseDate + time::Days{7 * weekIdx} + time::Hours{hour} +
                    time::Minutes{minute};

    if (!typologies::appendBoundedTxn(
            ctx, out, budget,
            transactions::Draft{
                .source = src,
                .destination = dst,
                .amount = amount,
                .timestamp = time::toEpochSeconds(ts),
                .isFraud = 1,
                .ringId = static_cast<std::int32_t>(plan.ringId),
                .channel = channel,
            })) {
      break;
    }
  }

  return out;
}

} // namespace PhantomLedger::transfers::fraud::typologies::invoice
